File: Kotlin/subarray-product-less-than-k.kt
class Solution {
    fun numSubarrayProductLessThanK(nums: IntArray, k: Int): Int {
        // no point checking product below zero
        if (k == 0) {
            return 0
        }
        var count = 0
        var prod = 1
        var left = 0
        // move the right index while the value is smaller than k
        for (right in nums.indices) {
            prod *= nums[right]
            // if product grows larger, shrink the left side
            while (left <= right && prod >= k) {
                prod /= nums[left++]
            }
            // accumulate on count the difference between left and right pointers (multiple sub-arrays)
            count += right - left + 1
        }
        return count
    }
}

File: Go/subarray-product-less-than-k.go
func numSubarrayProductLessThanK(nums []int, k int) int {
    
    if k == 0 {
        return 0
    }

    prod, count, left := 1, 0, 0

    for right := range nums {
        prod *= nums[right]
        for left <= right && prod >= k {
            prod /= nums[left]
            left++
        }
        count += right - left + 1
    }

    return count
    
}